File: LV 2/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    answer = 0
    idx = 0
    cur_weight = 0
    time = 0
    queue = deque()
    while True:
        if queue and queue[0][1] <= time:
            cur_weight = cur_weight - queue[0][0]
            queue.popleft()
            
        if idx < len(truck_weights) and weight >= cur_weight + truck_weights[idx]: 
            queue.append([truck_weights[idx], time+bridge_length])
            cur_weight = cur_weight + truck_weights[idx]
            idx += 1
            
        time += 1
        if idx == len(truck_weights) and not queue:
            break
        
    answer = time
    return answer

File: LV 2/다리를 지나는 트럭.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {

    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int time = 1;
        Queue<Integer> truckQueue = new LinkedList<Integer>();
        Queue<Integer> timeQueue = new LinkedList<Integer>();
        int currentBridgeWeight = 0;
        int truckIndex = 0;
        while (truckIndex < truck_weights.length) {
            if (!truckQueue.isEmpty()) {
                if (time == timeQueue.peek()) {
                    timeQueue.poll();
                    currentBridgeWeight -= truckQueue.poll();
                }
            }
            
            if (currentBridgeWeight + truck_weights[truckIndex] <= weight) {
                currentBridgeWeight += truck_weights[truckIndex];
                truckQueue.offer(truck_weights[truckIndex++]);
                timeQueue.offer(time + bridge_length);
            }
            
            time++;
        }
        return timeQueue.stream().mapToInt(i -> i).max().getAsInt();
    }
}
